beacon-chain/state/state-native/v1/state_trie.go

```diff
@@ -72,46 +112,89 @@ func InitializeFromProtoUnsafe(st *ethpb.BeaconState) (*BeaconState, error) {
+	b.sharedFieldReferences[blockRoots] = stateutil.NewRef(1)
+	b.sharedFieldReferences[previousEpochAttestations] = stateutil.NewRef(1)
+	b.sharedFieldReferences[currentEpochAttestations] = stateutil.NewRef(1)
+	b.sharedFieldReferences[slashings] = stateutil.NewRef(1)
+	b.sharedFieldReferences[eth1DataVotes] = stateutil.NewRef(1)
+	b.sharedFieldReferences[validators] = stateutil.NewRef(1)
+	b.sharedFieldReferences[balances] = stateutil.NewRef(1)
+	b.sharedFieldReferences[historicalRoots] = stateutil.NewRef(1)
+
+	state.StateCount.Inc()
+	return b, nil
+}
+
+// Copy returns a deep copy of the beacon state.
+func (b *BeaconState) Copy() state.BeaconState {
 	b.lock.RLock()
 	defer b.lock.RUnlock()
 	fieldCount := params.BeaconConfig().BeaconStateFieldCount
 	dst := &BeaconState{
-		state: &ethpb.BeaconState{
-			// Primitive types, safe to copy.
-			GenesisTime:      b.state.GenesisTime,
-			Slot:             b.state.Slot,
-			Eth1DepositIndex: b.state.Eth1DepositIndex,
-
-			// Large arrays, infrequently changed, constant size.
-			RandaoMixes:               b.state.RandaoMixes,
-			StateRoots:                b.state.StateRoots,
-			BlockRoots:                b.state.BlockRoots,
-			PreviousEpochAttestations: b.state.PreviousEpochAttestations,
-			CurrentEpochAttestations:  b.state.CurrentEpochAttestations,
-			Slashings:                 b.state.Slashings,
-			Eth1DataVotes:             b.state.Eth1DataVotes,
-
-			// Large arrays, increases over time.
-			Validators:      b.state.Validators,
-			Balances:        b.state.Balances,
-			HistoricalRoots: b.state.HistoricalRoots,
-
-			// Everything else, too small to be concerned about, constant size.
-			Fork:                        b.fork(),
-			LatestBlockHeader:           b.latestBlockHeader(),
-			Eth1Data:                    b.eth1Data(),
-			JustificationBits:           b.justificationBits(),
-			PreviousJustifiedCheckpoint: b.previousJustifiedCheckpoint(),
-			CurrentJustifiedCheckpoint:  b.currentJustifiedCheckpoint(),
-			FinalizedCheckpoint:         b.finalizedCheckpoint(),
-			GenesisValidatorsRoot:       b.genesisValidatorRoot(),
-		},
+		// Primitive types, safe to copy.
+		genesisTime:      b.genesisTime,
+		slot:             b.slot,
+		eth1DepositIndex: b.eth1DepositIndex,
+
+		// Large arrays, infrequently changed, constant size.
+		slashings: b.slashings,
+
+		// Large arrays, infrequently changed, constant size.
+		blockRoots:                b.blockRoots,
+		stateRoots:                b.stateRoots,
+		randaoMixes:               b.randaoMixes,
+		previousEpochAttestations: b.previousEpochAttestations,
+		currentEpochAttestations:  b.currentEpochAttestations,
+		eth1DataVotes:             b.eth1DataVotes,
+
+		// Large arrays, increases over time.
+		balances:        b.balances,
+		historicalRoots: b.historicalRoots,
+		validators:      b.validators,
+
+		// Everything else, too small to be concerned about, constant size.
+		genesisValidatorsRoot: b.genesisValidatorsRoot,
```

Comment:
You are changing existing behaviour for the below fields. Previously it was a full copy, but now you are just copying its reference

Reply:
the underlying type is `[32]byte`, so it is actually a full copy, no?

Reply:
Referring to all these fields:
```
		justificationBits:     b.justificationBits,

		// Everything else, too small to be concerned about, constant size.
		fork:                        b.fork,
		latestBlockHeader:           b.latestBlockHeader,
		eth1Data:                    b.eth1Data,
		previousJustifiedCheckpoint: b.previousJustifiedCheckpoint,
		currentJustifiedCheckpoint:  b.currentJustifiedCheckpoint,
		finalizedCheckpoint:         b.finalizedCheckpoint,
```